nixos/modules/services/web-apps/baget.nix

```diff
@@ -0,0 +1,104 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.baget;
+
+  defaultConfig = {
+    "PackageDeletionBehavior" = "Unlist";
+    "AllowPackageOverwrites" = false;
+
+    "Database" = {
+      "Type" = "Sqlite";
+      "ConnectionString" = "Data Source=baget.db";
+    };
+
+    "Storage" = {
+      "Type" = "FileSystem";
+      "Path" = "";
+    };
+
+    "Search" = {
+      "Type" = "Database";
+    };
+
+    "Mirror" = {
+      "Enabled" = false;
+      "PackageSource" = "https://api.nuget.org/v3/index.json";
+    };
+
+    "Logging" = {
+      "IncludeScopes" = false;
+      "Debug" = {
+        "LogLevel" = {
+          "Default" = "Warning";
+        };
+      };
+      "Console" = {
+        "LogLevel" = {
+          "Microsoft.Hosting.Lifetime" = "Information";
+          "Default" = "Warning";
+        };
+      };
+    };
+  };
+
+  configAttrs = recursiveUpdate defaultConfig cfg.extraConfig // { "ApiKey" = "REPLACE_WITH_API_KEY"; };
+
+  configFile = pkgs.writeText "appsettings.json" (builtins.toJSON configAttrs);
+
+in
+{
+  options.services.baget = {
+    enable = mkEnableOption "BaGet web application.";
+
+    apiKeyFile = mkOption {
+      type = types.path;
+      example = "/root/baget.key";
+      description = ''
+        Private API key for BaGet.
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.attrs;
+      default = {};
+      example = {
+        "Database" = {
+          "Type" = "PostgreSql";
+          "ConnectionString" = "Server=/run/postgresql;Port=5432;";
+        };
+      };
+      description = ''
+        Extra configuration options for BaGet. Refer to <link xlink:href="https://loic-sharma.github.io/BaGet/configuration/"/> for details.
+      '';
+    };
+  };
+
+  # implementation
+
+  config = mkIf cfg.enable {
+
+    systemd.services.baget = {
+      description = "BaGet server";
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ];
+      after = [ "network.target" "network-online.target" ];
+      serviceConfig = {
+        ExecStart = "${pkgs.baget}/bin/BaGet";
+        WorkingDirectory = "/var/lib/baget";
+        DynamicUser = true;
+        PermissionsStartOnly = true;
+        StateDirectory = "baget";
+        StateDirectoryMode = "0700";
+      };
+      preStart = ''
+        sed "s,REPLACE_WITH_API_KEY,$(cat ${cfg.apiKeyFile})," ${configFile} > /var/lib/baget/appsettings.json
```

Comment:
this feels pretty fragile. would it make sense to use the systemd credentials mechanism and jq to merge the generated config and the api key fragment?

Reply:
Didn't know it exists! I'll try it out.